#2.14 improve terrain

Manfred-Hastmark opened the pull request:
Improved noise done

Implemented noise layers with settings:

![image](https://user-images.githubusercontent.com/89832008/226685408-ed49a9ff-d958-44a9-9832-09375c94cd80.png)

Improved on code in chunkhandler and chunk. There's now dynamic chunkresolutions in three stages, and empty chunks are also removed, furthermore chunks are created upon loading them instead of generating all chunks when arriving to the planet (startplanet excluded). Below are the settings for the different resolutions for different distances and also the chunksettings.

![image](https://user-images.githubusercontent.com/89832008/226686125-07b73fb5-4a96-4b2b-b0e0-95425f17368e.png)

Also doubled the size of the planets. 

Next to do with the terrain is to create biomes, then create terrain after the biomes, then create caves and overhangs.

Turbobus commented:
Looks and works much better than we had before and only a little less performance for double the size of the planets good job!

- One thing I noticed however is that you do not have the latest main in your branch so if you could merge before it would probably be good. 

- Also is it possible in some way to avoid the seems that are currently created when the resolution changes? the image below shows what seems I mean:
![image](https://user-images.githubusercontent.com/59963570/226795549-19491c13-340c-4ab0-82b6-bdcf8a9c624f.png)

- And lastly, is there a way to avoid the massive lagspike when leaving the planets now? If it is the whole planet loading in that takes time (the marching cubes algorithm) is it possible to load the mesh on the GPU in the background when moving away from the planet and then have it swap seemlessly.

Manfred-Hastmark commented:
> Looks and works much better than we had before and only a little less performance for double the size of the planets good job!
> 
> * One thing I noticed however is that you do not have the latest main in your branch so if you could merge before it would probably be good.
> * Also is it possible in some way to avoid the seems that are currently created when the resolution changes? the image below shows what seems I mean:
>   ![image](https://user-images.githubusercontent.com/59963570/226795549-19491c13-340c-4ab0-82b6-bdcf8a9c624f.png)
> * And lastly, is there a way to avoid the massive lagspike when leaving the planets now? If it is the whole planet loading in that takes time (the marching cubes algorithm) is it possible to load the mesh on the GPU in the background when moving away from the planet and then have it swap seemlessly.


* I have merged the last changes on main into my branch now

* Regarding the seems you are talking about, I will probably not be able to fix these, maybe they can be fixed by having more chunks loaded in high res and adding fog to hide the seems. I can add some adjustable variables for the ranges in which the different resolutions should be loaded and then we can hide the seems later with fog or something.

* About the lagspikes, I will start working on this, will get back when I'm done.

* Also encountered a new bug when merging main into my branch, the player is now able to jump extremely high up, this has to be addressed. I don't know if this is a work for me or someone else? It has probably occured when @Danilll01 changed so that we use radius instead of diameter, I know I had this issue when I was trying to do the same thing, hence why I didn't do it immediately, but left if for the users of planet.diameter to change. Though I could take a look if this is considered to be an issue with my code. The current code on this branch has this bug if anyone is interested in investigating.

Manfred-Hastmark commented:
> 

* I have now removed the lagspike, I fixed it by generating highres and lowres mesh for all planets on startup and then dissabling and enabling them accordingly. slight FPS drop to due a lot more transforms to move and a lot more memory usage. Still ok though, get around 160-350 FPS with 10 planets in build version, uses ~2.5GB RAM when running (build version), loading times have increased though, still not that bad considering.

* Also added option to change the radius in which the different resolutions should be, thus the person later on making the fog to hide the seems can determine how big the highres, mediumres and lowres areas should be to be able to hide the seems nicely. (Chunk handler, planet prefab)
![image](https://user-images.githubusercontent.com/89832008/227056196-44499c38-d770-494e-8c53-da525bfabe24.png)

* Furthermore added variable which states how many chunks should be checked each frame, culling, changing res etc. done in this update of the chunk. Also under chunk handler in planet prefab, can be played around with when we get to Quality settings maybe as some computers will be able to handle a higher number here.
![image](https://user-images.githubusercontent.com/89832008/227056920-fee0c557-05fb-45f7-88c0-4732497cf216.png)

* Player jumping issue still persists, not sure if I should fix it or not, could maybe take a look at it tomorrow.

Turbobus commented:
So should this be merged or not right now?

Manfred-Hastmark commented:
> 

I will take a look at the player movement bug

Manfred-Hastmark commented:
> > 
> 
> I will take a look at the player movement bug

Also seems like the moons are getting to close to the planets as well, I'm getting sucked up into the moons when they are passing by sometimes, even if I'm not jumping

Manfred-Hastmark commented:
I found a fix for the player movement bug, it seems that the mass was incorrect after changing to radius. As radius = 2 * diameter => diameter * diameter = 4 * radius * radius, not radius * radius, see below. Correct me if this fix is wrong.

The issue with moons geting to close to planets still persists though, I will take a look at this as well.

### Before

```cs
/// <summary>
/// Set up the values for the planets
/// </summary>
public void SetUpPlanetValues()
{
    mass = surfaceGravity * diameter * diameter / Universe.gravitationalConstant;
    gameObject.name = bodyName;
}
```
### After Radius change
```cs
/// <summary>
/// Set up the values for the planets
/// </summary>
public void SetUpPlanetValues()
{
    mass = surfaceGravity * radius * radius / Universe.gravitationalConstant;
    gameObject.name = bodyName;
}
```
### Now
```cs
/// <summary>
/// Set up the values for the planets
/// </summary>
public void SetUpPlanetValues()
{
    mass = surfaceGravity * 4 * radius * radius / Universe.gravitationalConstant;
    gameObject.name = bodyName;
}
````

Manfred-Hastmark commented:
> > > 
> > 
> > 
> > I will take a look at the player movement bug
> 
> Also seems like the moons are getting to close to the planets as well, I'm getting sucked up into the moons when they are passing by sometimes, even if I'm not jumping

Think I found the issue. Assume i = 1, then the moon will spawn on the edge of the radius, meaning it will get close. Also we are using Unity random which is not correct. Anyways, I will go ahead and fix this, also it could maybe be nice to randomize the distances the moons have from the planet as they are just a set length from the planet rn. I will leave this for someone else though.

```cs
// Instantiate moons for the given planet
private void InstantiateMoons(Planet parentPlanet, int numberOfMoons)
{
    for (int i = 1; i < numberOfMoons + 1; i++)
    {
        GameObject moon = Instantiate(planetsPrefab);
        moon.transform.parent = parentPlanet.transform;
        moon.transform.localPosition = RandomPointOnCircleEdge(parentPlanet.radius * i); //For i = 1, moon orbits on planetRadius
        moon.gameObject.name = "Moon " + i;

        Planet moonBody = moon.GetComponent<Planet>();
        moonBody.bodyName = "Moon " + i;
        moonBody.radius = random.Next((int)(parentPlanet.radius / 5), (int)((parentPlanet.radius / 2) + 1));
        moonBody.SetUpPlanetValues();
        moonBody.Initialize(player.transform, random.Next(), false); //False here beacause we don't spawn on moons
        parentPlanet.moons.Add(moonBody);
        SetupOrbitComponents(parentPlanet.gameObject, moon);
    }
}

// Gives back a random position on the edge of a circle given the radius of the circle
private Vector3 RandomPointOnCircleEdge(float radius)
{
    var vector2 = Random.insideUnitCircle.normalized * radius; //Unity random and always the same distance
    return new Vector3(vector2.x, 0, vector2.y);
}
```

Manfred-Hastmark commented:
> > > > 
> > > 
> > > 
> > > I will take a look at the player movement bug
> > 
> > 
> > Also seems like the moons are getting to close to the planets as well, I'm getting sucked up into the moons when they are passing by sometimes, even if I'm not jumping
> 
> Think I found the issue. Assume i = 1, then the moon will spawn on the edge of the radius, meaning it will get close. Also we are using Unity random which is not correct. Anyways, I will go ahead and fix this, also it could maybe be nice to randomize the distances the moons have from the planet as they are just a set length from the planet rn. I will leave this for someone else though.
> 
> ```cs
> // Instantiate moons for the given planet
> private void InstantiateMoons(Planet parentPlanet, int numberOfMoons)
> {
>     for (int i = 1; i < numberOfMoons + 1; i++)
>     {
>         GameObject moon = Instantiate(planetsPrefab);
>         moon.transform.parent = parentPlanet.transform;
>         moon.transform.localPosition = RandomPointOnCircleEdge(parentPlanet.radius * i); //For i = 1, moon orbits on planetRadius
>         moon.gameObject.name = "Moon " + i;
> 
>         Planet moonBody = moon.GetComponent<Planet>();
>         moonBody.bodyName = "Moon " + i;
>         moonBody.radius = random.Next((int)(parentPlanet.radius / 5), (int)((parentPlanet.radius / 2) + 1));
>         moonBody.SetUpPlanetValues();
>         moonBody.Initialize(player.transform, random.Next(), false); //False here beacause we don't spawn on moons
>         parentPlanet.moons.Add(moonBody);
>         SetupOrbitComponents(parentPlanet.gameObject, moon);
>     }
> }
> 
> // Gives back a random position on the edge of a circle given the radius of the circle
> private Vector3 RandomPointOnCircleEdge(float radius)
> {
>     var vector2 = Random.insideUnitCircle.normalized * radius; //Unity random and always the same distance
>     return new Vector3(vector2.x, 0, vector2.y);
> }
> ```

Moons are now further away

![image](https://user-images.githubusercontent.com/89832008/227171798-8e226788-3b91-40ce-b3ca-50242292a2d1.png)

```cs
// Instantiate moons for the given planet
private void InstantiateMoons(Planet parentPlanet, int numberOfMoons)
{
    for (int i = 1; i < numberOfMoons + 1; i++)
    {
        GameObject moon = Instantiate(planetsPrefab);
        moon.transform.parent = parentPlanet.transform;
        moon.transform.localPosition = RandomPointOnCircleEdge(parentPlanet.radius * (i + 1));
        moon.gameObject.name = "Moon " + i;

        Planet moonBody = moon.GetComponent<Planet>();
        moonBody.bodyName = "Moon " + i;
        moonBody.radius = random.Next((int)(parentPlanet.radius / 5), (int)((parentPlanet.radius / 2) + 1));
        moonBody.SetUpPlanetValues();
        moonBody.Initialize(player.transform, random.Next(), false); //False here beacause we don't spawn on moons
        parentPlanet.moons.Add(moonBody);
        SetupOrbitComponents(parentPlanet.gameObject, moon);
    }
}

// Gives back a random position on the edge of a circle given the radius of the circle
private Vector3 RandomPointOnCircleEdge(float radius)
{
    var vector2 = random.OnUnitCircle() * radius;
    return new Vector3(vector2.x, 0, vector2.y);
}
```
